File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        ///base condition
        return dfs(0, root);
    }
    private int dfs(int dep, TreeNode root) {
        if(root == null) {
            return dep;
        }
        int left = dfs(dep+1, root.left);
        int right = dfs(dep+1, root.right);
        return Math.max(left, right);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return dfs(root) != -1;
    }
    int dfs(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        int leftt=dfs(root->left);
        if(leftt == -1) return -1;
        int rightt =dfs(root->right);
        if(rightt ==-1) return -1;

        if(abs(leftt-rightt)>1) return -1;
        return max(leftt,rightt)+1;
    }
};